File: 0017 Letter Combinations of a Phone Number/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string chars[8] = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    
    vector<string> letterCombinations(string digits) {
        if (digits.empty())  return vector<string>();
        
        vector<string> state(1, "");
        for (auto u : digits)
        {
            vector<string> now;
            for (auto c : chars[u - '2'])
                for (auto s : state)
                    now.push_back(s + c);
            state = now;
        }
        return state;
    }
};

int main(){
    Solution s;
    for (const auto &str : s.letterCombinations("23"))
        cout << str << "\t";

    return 0;
}

File: 0017 Letter Combinations of a Phone Number/solution.py
#! python3

class Solution:
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if not digits:
            return []
        
        self.dict = {"2":"abc", "3":"def", "4":"ghi", "5":"jkl", "6":"mno", "7":"pqrs","8":"tuv","9":"wxyz",}
        res = ['']
        for digit in digits:
            res = [a + b for a in res for b in self.dict[digit]]
                
        return res

if __name__ == "__main__":
    print(Solution().letterCombinations("23"))